File: server/go/pkg/handlers/colorize.go
package handlers

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/cstrab/colorizer/server/go/pkg/log"
	"github.com/cstrab/colorizer/server/go/pkg/types"
	"github.com/cstrab/colorizer/server/go/pkg/utils"
)

func setCorsHeaders(w http.ResponseWriter) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Expose-Headers", "Content-Disposition")
}

func ColorizeImages(w http.ResponseWriter, r *http.Request, numWorkers int) {
	setCorsHeaders(w)
	logger := log.GetLogger()
	logger.Info("%s - \"%s %s %s\"", r.RemoteAddr, r.Method, r.URL.Path, r.Proto)

	file, header, err := r.FormFile("file")
	if err != nil {
		logger.Error("Failed to retrieve file: %v", err)
		http.Error(w, "Failed to retrieve file", http.StatusBadRequest)
		return
	}
	defer file.Close()

	if filepath.Ext(header.Filename) != ".zip" {
		logger.Error("Invalid file format. Expected .zip, got %s", filepath.Ext(header.Filename))
		http.Error(w, "Invalid file format. Please upload a .zip file.", http.StatusBadRequest)
		return
	}

	colorExtension := r.FormValue("color_extension")
	if colorExtension == "" {
		logger.Error("Color extension is required but not provided.")
		http.Error(w, "Color extension is required.", http.StatusBadRequest)
		return
	}

	colorMappingJSON := r.FormValue("color_mapping")
	if colorMappingJSON == "" {
		logger.Error("Color mapping is required but not provided.")
		http.Error(w, "Color mapping is required.", http.StatusBadRequest)
		return
	}

	var colorMappings []types.ColorMapping
	if err := json.Unmarshal([]byte(colorMappingJSON), &colorMappings); err != nil {
		logger.Error("Failed to parse color mapping: %v", err)
		http.Error(w, "Invalid color mapping format.", http.StatusBadRequest)
		return
	}

	outputZipBuffer, err := utils.ProcessImages(file, header.Size, colorExtension, colorMappings, numWorkers)
	if err != nil {
		logger.Error("Failed to process images: %v", err)
		http.Error(w, "Failed to process images", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/zip")
	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s%s.zip\"", strings.TrimSuffix(header.Filename, filepath.Ext(header.Filename)), colorExtension))

	if _, err := io.Copy(w, outputZipBuffer); err != nil {
		logger.Error("Failed to stream response: %v", err)
		http.Error(w, "Failed to stream response", http.StatusInternalServerError)
		return
	}

	logger.Info("%s - \"%s %s %s\" %d %s", r.RemoteAddr, r.Method, r.URL.Path, r.Proto, http.StatusOK, http.StatusText(http.StatusOK))
}

File: server/python/handlers/colorize.py
import json
from fastapi import File, UploadFile, HTTPException, Form, Request
from fastapi.responses import StreamingResponse

from utils.process import process_images, iter_file
from log.log import get_logger
from models.models import ColorMapping

logger = get_logger()

async def colorize_images(
    request: Request,
    file: UploadFile = File(...),
    color_extension: str = Form(...),
    color_mapping: str = Form(...)
):
    num_workers = request.app.state.num_workers
    
    if not file.filename.endswith('.zip'):
        logger.error(f"Invalid file format: {file.filename}")
        raise HTTPException(status_code=400, detail="Invalid file format. Please upload a .zip file.")
    if not color_extension:
        logger.error("Color extension is required.")
        raise HTTPException(status_code=400, detail="Color extension is required.")
    if not color_mapping:
        logger.error("Color mapping is required.")
        raise HTTPException(status_code=400, detail="Color mapping is required.")

    try:
        color_mapping = [ColorMapping(**cm) for cm in json.loads(color_mapping)]
    except Exception as e:
        logger.error(f"Invalid color mapping format: {str(e)}")
        raise HTTPException(status_code=400, detail="Invalid color mapping format.")

    try:
        zip_bytes = await file.read()
        output_zip_bytes = await process_images(zip_bytes, color_extension, color_mapping, num_workers)

        if output_zip_bytes is None:
            logger.error("Image processing failed.")
            raise HTTPException(status_code=500, detail="Image processing failed.")

        output_zip_bytes.seek(0)
        original_filename = file.filename.rsplit('.', 1)[0]

        return StreamingResponse(iter_file(output_zip_bytes), media_type="application/zip", headers={
            "Content-Disposition": f"attachment; filename=\"{original_filename}{color_extension}.zip\""
        })
    except Exception as e:
        logger.error(f"Failed to process file: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to process file: {str(e)}")